File: frontend/src/api/notifications.ts
// Copyright (c) 2026, AssetCore Team
//
// Notification Framework (Wave N1) — API client cho per-user email toggle.
// Spec: docs/imm-00/05_API_Specification.md §III.21
// Mirrors BE: assetcore.api.notifications.{get_notification_preferences, set_email_enabled}
//
// In-app (chuông) dùng Frappe core desk Notification Log — không có client riêng ở đây.
import { frappeGet, frappePost } from './helpers'

const BASE = '/api/method/assetcore.api.notifications'

/** Tùy chọn thông báo của user — khớp BE envelope data. */
export interface NotificationPreferences {
  email_enabled: boolean
}

/** GET get_notification_preferences. frappeGet đã unwrap envelope → trả T trực tiếp. */
export function getNotificationPreferences(): Promise<NotificationPreferences> {
  return frappeGet(`${BASE}.get_notification_preferences`)
}

/** POST set_email_enabled — bật/tắt nhận email cho user hiện tại. */
export function setEmailEnabled(enabled: boolean): Promise<NotificationPreferences> {
  return frappePost(`${BASE}.set_email_enabled`, { enabled })
}

/** Ngưỡng màu KPI (khớp BE _delivery_status / _opt_out_status). */
export type KpiStatus = 'good' | 'warn' | 'bad' | 'na'

/**
 * KPI độ phủ thông báo (System Manager only) — khớp BE envelope data.
 * Spec: docs/imm-00/04_Backend_Design.md §III.1b-4.
 */
export interface DeliveryKpi {
  delivery_rate: number | null // null = mẫu rỗng (chia-0 guard)
  sent: number
  failed: number
  opt_out_rate: number | null // null = không có user
  total_users: number
  opted_out: number
  window_days: number
  delivery_status: KpiStatus
  opt_out_status: KpiStatus
}

/** GET get_delivery_kpi — KPI delivery rate + opt-out rate (cửa sổ `days` ngày). */
export function getDeliveryKpi(days = 30): Promise<DeliveryKpi> {
  return frappeGet(`${BASE}.get_delivery_kpi`, { days })
}

File: assetcore/api/notifications.py
# Copyright (c) 2026, AssetCore Team
# Notification Framework (Wave N1) — Tier 1 API Layer.
#
# Endpoint cho per-user email toggle. In-app (chuông) dùng API Frappe core sẵn
# có (Notification Log) — không cần endpoint riêng ở đây.
#
# Spec: docs/imm-00/05_API_Specification.md §III.21.

from __future__ import annotations

import frappe

from assetcore.services import notifications as svc
from assetcore.utils.api_handler import handle


@frappe.whitelist()
def get_notification_preferences(user: str = "") -> dict:
    """Đọc tùy chọn nhận email của user hiện tại (System Manager có thể truyền user).

    Returns envelope: {"success": true, "data": {"email_enabled": bool}}
    """
    return handle(svc.get_notification_preferences, user or None)


@frappe.whitelist(methods=["POST"])
def set_email_enabled(enabled: bool = True, user: str = "") -> dict:
    """Bật/tắt nhận email cho user hiện tại.

    Body: {"enabled": false}. Returns: {"success": true, "data": {"email_enabled": bool}}
    """
    # Frappe parse JSON body bool đúng kiểu; cast phòng trường hợp truyền string.
    if isinstance(enabled, str):
        enabled = enabled.strip().lower() in ("1", "true", "yes")
    return handle(svc.set_email_enabled, bool(enabled), user or None)


@frappe.whitelist()
def get_delivery_kpi(days: str = "30") -> dict:
    """KPI độ phủ thông báo (System Manager only): delivery rate + opt-out rate.

    Query: `days` (cửa sổ Email Queue, mặc định 30). Cast str→int an toàn (tránh
    HTTP 417 do type-hint `int` trên GET param — LL-BE-1).

    Returns envelope: {"success": true, "data": {delivery_rate, opt_out_rate, ...}}.
    Spec: docs/imm-00/05_API_Specification.md §III.21.
    """
    try:
        days_int = int(days)
    except (TypeError, ValueError):
        days_int = 30
    return handle(svc.get_delivery_kpi, days_int)
